datastore.go

```diff
@@ -167,13 +190,14 @@ func (d *datastore) Close() error {
 func (d *datastore) IsThreadSafe() {}
 
 type badgerBatch struct {
-	entries []*badger.Entry
-	db      *badger.KV
+	db  *badger.DB
+	txn *badger.Txn
 }
 
 func (d *datastore) Batch() (ds.Batch, error) {
 	return &badgerBatch{
-		db: d.DB,
+		db:  d.DB,
+		txn: d.DB.NewTransaction(true),
```

Comment:
According to docs 'it is absolutely essential to call Discard()' on a transaction, this is done internally by Commit. 

There may be cases where if `Put`/`Get` return an error, `Commit` may not get called. If that happens we either need to make sure that Commit gets called, or add `Abort` method to datastore.Batch interface.

Reply:
I assume they mean "absolutely essential" for performance/memory. We should probably have an abort but this also looks like a good place for a finalizer.

Reply:
We could also poison the batch on error and abort automatically (I'm guessing that's what users will usually do anyways).

Reply:
Looking at current uses of batch in go-ipfs this should work

Reply:
Done in 1.0.2